File: scala/digital-wallet/src/main/scala/adapters/InvestmentPolicyInMemoryDatabase.scala
package adapters

import core.domain.entities.InvestmentPolicy
import core.domain.model.CreateInvestmentPolicyRequest
import ports.InvestmentPolicyDatabase

import java.util.UUID
import scala.collection.mutable

class InvestmentPolicyInMemoryDatabase extends InvestmentPolicyDatabase {
  val investmentPolicies: mutable.Map[String, InvestmentPolicy] = mutable.Map()

  override def insert(request: CreateInvestmentPolicyRequest): InvestmentPolicy = {
    val id = UUID.randomUUID().toString

    val investmentPolicy = InvestmentPolicy(
      id = id,
      allocationStrategy = request.allocationStrategy
    )

    investmentPolicies(id) = investmentPolicy
    investmentPolicy
  }

  override def findById(id: String): Option[InvestmentPolicy] = investmentPolicies.get(id)

  def clear(): Unit = {
    investmentPolicies.clear()
  }
}

File: kotlin/digital-wallet/src/main/kotlin/digitalwallet/adapters/InvestmentPolicyInMemoryDatabase.kt
package digitalwallet.adapters

import digitalwallet.core.domain.entities.InvestmentPolicy
import digitalwallet.core.domain.models.CreateInvestmentPolicyRequest
import digitalwallet.ports.InvestmentPolicyDatabase
import java.util.UUID
import digitalwallet.core.domain.models.InvestmentPolicy as InvestmentPolicyModel

class InvestmentPolicyInMemoryDatabase : InvestmentPolicyDatabase {
    val investmentPolicies = mutableMapOf<String, InvestmentPolicy>()

    override fun insert(request: CreateInvestmentPolicyRequest): InvestmentPolicyModel {
        val id = UUID.randomUUID().toString()

        val investmentPolicy =
            InvestmentPolicy(
                id = id,
                allocationStrategy = request.allocationStrategy,
            )

        investmentPolicies[id] = investmentPolicy

        return investmentPolicy.toModel()
    }

    override fun findById(id: String): InvestmentPolicyModel? = investmentPolicies[id]?.toModel()

    fun clear() {
        investmentPolicies.clear()
    }
}
